### ISSUE TITLE
No way to maintain a11y focus with local keys if layout changes

### ISSUE THREAD

**Issue by edeetee** [flutter/flutter#158899]
### Steps to reproduce

1. Take the code below
2. Run in debug mode with screen reader on ios
3. Click on the "moving counter" button using the screenreader

### Expected results

Focus stays on the widget

### Actual results

the a11y focus is reset to the top of the app

### Code sample

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a purple toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  final Key semanticKey = Key("moving_button");

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    final movingButton = Semantics(
        key: semanticKey,
        child: ElevatedButton(
            onPressed: _incrementCounter, child: const Text('MOVING Counter')));

    final stationaryButton = Semantics(
        child: ElevatedButton(
            onPressed: _incrementCounter,
            child: const Text('STATIONARY Counter')));

    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
            Semantics(
              child: AnimatedSwitcher(
                duration: const Duration(milliseconds: 500),
                child: _counter % 2 == 0
                    ? movingButton
                    : Row(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          Text('Something else'),
                          movingButton,
                        ],
                      ),
              ),
            ),
            stationaryButton
          ],
        ),
      ),
    );
  }
}

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/user-attachments/assets/bb4a5de0-fe5d-4400-9183-ae83707cb13a

</details>

### Logs

<details open><summary>Logs</summary>

```console
Launching lib/main.dart on Edward’s iPhone in debug mode...
Automatically signing iOS for device deployment using specified development team in Xcode project: MZ7437V6KU
Xcode build done.                                           14.4s
You may be prompted to give access to control Xcode. Flutter uses Xcode to run your app. If access is not allowed, you can change this through your Settings > Privacy & Security > Automation.
Connecting to VM Service at ws://127.0.0.1:61641/eXC1sev2rz0=/ws
Connected to the VM Service.
Reloaded 1 of 709 libraries in 296ms.
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
➜ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.24.3, on macOS 15.0 24A335
    darwin-arm64, locale en-NZ)
[!] Android toolchain - develop for Android devices (Android SDK
    version 34.0.0)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for
      more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK
      licenses.
      See https://flutter.dev/to/macos-android-setup for more
      details.
[✓] Xcode - develop for iOS and macOS (Xcode 16.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2024.2)
[!] Android Studio (version unknown)
    ✗ Unable to determine Android Studio version.
    ✗ android-studio-dir = /
    ✗ Android Studio not found at /Contents
[✓] VS Code (version 1.95.2)
[✓] Connected device (4 available)
[✓] Network resources

! Doctor found issues in 2 categories.
```

</details>

**Comment by edeetee** [comment]
Somewhat related to https://github.com/flutter/flutter/issues/105510

**Comment by danagbemava-nc** [comment]
Hi @edeetee, can you check the master channel to see if you still experience this issue? I tested this on stable (it wasn't consistent for me) but on master it seemed like the behavior is far better. 

If you still experience this on the master channel, please provide the version of the master channel that you used. 

Thank you

**Comment by edeetee** [comment]
Verified as fixed as of 3.27.0-1.0.pre.50!
Seems consistently fixed to me! I'll leave it to you if you want to close this or not if you're still seeing unstable behavior.
Any idea which commit did it? I'll keep an eye out for when it is released.

Nice work whoever fixed it :)

**Comment by edeetee** [comment]
Okay actually not fixed. It's very strange though.
It works on initial launch, but if I put the app to the background and return to it, the bug returns.

**Comment by edeetee** [comment]
https://github.com/user-attachments/assets/aa86c77f-85e2-46ff-9c3a-fc8347b79923

**Comment by danagbemava-nc** [comment]
You're right. It took me quite a few tries but I was able to reproduce the issue on the master channel as well. I got this early in my reproduction on android, but it remained consistent after a while. 

I tested on my iPhone XS running iOS 18.1 and Pixel 7 running android 15.

| iOS | android | 
| -- | -- |
| <video src="https://github.com/user-attachments/assets/81aa7395-5a0f-4864-8a82-974ef6f11c0d" /> | <video src="https://github.com/user-attachments/assets/4ba74362-b0f7-4afe-a79d-a7729f205006" /> |

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.24.5, on macOS 15.0.1 24A348 darwin-arm64, locale en-US)
    • Flutter version 3.24.5 on channel stable at /Users/deanli/dev/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision dec2ee5c1f (35 hours ago), 2024-11-13 11:13:06 -0800
    • Engine revision a18df97ca5
    • Dart version 3.5.4
    • DevTools version 2.37.3
```

```
[!] Flutter (Channel master, 3.27.0-1.0.pre.511, on macOS 15.0.1 24A348 darwin-arm64, locale en-US)
    • Flutter version 3.27.0-1.0.pre.511 on channel master at /Users/deanli/dev/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision eaa99f28b1 (5 hours ago), 2024-11-14 17:38:26 -0800
    • Engine revision 619804c0fb
    • Dart version 3.7.0 (build 3.7.0-140.0.dev)
    • DevTools version 2.41.0-dev.2
```

</details>

**Comment by chunhtai** [comment]
In general, we should avoid moving the control when activate, but I think you may be able to fix this by doing something like
```dart
Semantics(
              child: AnimatedSwitcher(
                duration: const Duration(milliseconds: 500),
                child: Row(
                        mainAxisSize: MainAxisSize.min,
                        children: [
                          if(_counter % 2 == 0)
                            Text('Something else'),
                          movingButton,
                        ],
                      ),
              ),
            ),
```

cc @hannah-hyj since you were working on focus

**Comment by edeetee** [comment]
I agree, this a somewhat contrived example. In our code, the widget is visually in the same place, its just the widget tree structure that is different.

**Comment by chunhtai** [comment]
> > In general, we should avoid moving the control when activate, but I think you may be able to fix this by doing something like
> > ```dart
> > Semantics(
> >               child: AnimatedSwitcher(
> >                 duration: const Duration(milliseconds: 500),
> >                 child: Row(
> >                         mainAxisSize: MainAxisSize.min,
> >                         children: [
> >                           if(_counter % 2 == 0)
> >                             Text('Something else'),
> >                           movingButton,
> >                         ],
> >                       ),
> >               ),
> >             ),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > cc @hannah-hyj since you were working on focus
> 
> Also, this fix will not work for my needs. I need to change the whole widget tree, not just a single widget sibling to my focused semantics.
> 
> Maybe there is just no way to do what I want? To be able to update the layout of the app while keeping the semantics focused on a given element?

If you can't maintain the structure, the other way will be using globalkey to maintain the state of the button

**Comment by edeetee** [comment]
> > > In general, we should avoid moving the control when activate, but I think you may be able to fix this by doing something like
> > > ```dart
> > > Semantics(
> > >               child: AnimatedSwitcher(
> > >                 duration: const Duration(milliseconds: 500),
> > >                 child: Row(
> > >                         mainAxisSize: MainAxisSize.min,
> > >                         children: [
> > >                           if(_counter % 2 == 0)
> > >                             Text('Something else'),
> > >                           movingButton,
> > >                         ],
> > >                       ),
> > >               ),
> > >             ),
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > cc @hannah-hyj since you were working on focus
> > 
> > 
> > Also, this fix will not work for my needs. I need to change the whole widget tree, not just a single widget sibling to my focused semantics.
> > Maybe there is just no way to do what I want? To be able to update the layout of the app while keeping the semantics focused on a given element?
> 
> If you can't maintain the structure, the other way will be using globalkey to maintain the state of the button

As mentioned in https://github.com/flutter/flutter/issues/105510, GlobalKeys can't be used inside AnimatedSwitchers, as during the animation it contains a copy of each.
